dropshot/src/handler.rs

```diff
@@ -262,7 +265,89 @@ where
         &self,
         rqctx: RequestContext<Context>,
         p: FuncParams,
-    ) -> HttpHandlerResult;
+    ) -> Result<Response<Body>, HandlerError>;
+}
+
+/// Errors returned by an endpoint handler function.
+///
+/// A function implementing an HTTP endpoint must return a `Result` with this
+/// error type. Typically, user code will only interact with this type via its
+/// `From` implementation, which allows it to be constructed from any value
+/// implementing the [`HttpResponseError`] trait. This type is largely an
+/// implementation detail, which is necessary to represent errors which may be
+/// errors returned by the endpoint's handler function *or* generated internally
+/// (i.e. by extractors).
+pub enum HandlerError {
+    Handler { message: String, rsp: Response<Body> },
+    Dropshot(HttpError),
+}
+
+impl HandlerError {
+    pub(crate) fn status_code(&self) -> StatusCode {
+        match self {
+            Self::Handler { ref rsp, .. } => rsp.status(),
+            Self::Dropshot(ref e) => e.status_code.as_status(),
+        }
+    }
+
+    pub(crate) fn internal_message(&self) -> &String {
+        match self {
+            Self::Handler { ref message, .. } => message,
+            Self::Dropshot(ref e) => &e.internal_message,
+        }
+    }
+
+    pub(crate) fn external_message(&self) -> Option<&String> {
+        match self {
+            Self::Handler { .. } => None,
+            Self::Dropshot(ref e) => Some(&e.internal_message),
+        }
+    }
+
+    pub(crate) fn into_response(self, request_id: &str) -> Response<Body> {
+        match self {
+            Self::Handler { rsp, .. } => rsp,
+            Self::Dropshot(e) => e.into_response(request_id),
+        }
+    }
+}
+
+impl<E> From<E> for HandlerError
+where
+    E: HttpResponseError,
+{
+    fn from(e: E) -> Self {
+        let message = e.to_string();
+        let status = e.status_code();
+        match e.to_response(Response::builder().status(status.as_status())) {
+            Ok(rsp) => Self::Handler { message, rsp },
+            Err(e) => Self::Dropshot(e),
+        }
+    }
+}
+
+/// An error type which can be converted into an HTTP response.
+///
+/// If a handler function's return value is a [`Result`], the error type must
+/// implement this trait, so that a response can be generated when the handler
+/// returns an error.  In order to implement this trait, a type must:
+///
+/// 1. Implement the [`HttpResponseContent`] trait, defining the content of the
+///    error's response body.  This trait is implemented for all types that
+///   implement the [`Serialize`] and [`JsonSchema`] traits.
```

Comment:
```suggestion
///    implement the [`Serialize`] and [`JsonSchema`] traits.
```

Or perhaps: this is most simply done by implementing the `Serialize` and `JsonSchema` traits

and maybe: ...for which there's a blanket implementation